src/lib/loaders/batchLoader.ts

```diff
@@ -0,0 +1,111 @@
+import DataLoader from "dataloader"
+import { chain, flatten, chunk } from "lodash"
+import config from "config"
+
+const { ENABLE_RESOLVER_BATCHING } = config
+
+/**
+ * This is just used to compare options sent into the dataloader.
+ * If the options are an object this stingifies the object in a way
+ * that they can be grouped as similar.
+ */
+const renderParams = key => {
+  if (typeof key === "string") {
+    return ""
+  }
+  const { id, ...params } = key
+  return Object.entries(params)
+    .map(entry => entry.join("="))
+    .sort()
+    .join("&")
+}
+
+interface GroupKeysResult {
+  id: string[]
+  size: number
+  [key: string]: any
+}
+export const groupKeys = (requestedKeys: string | { id }): GroupKeysResult[] =>
+  (chain(requestedKeys)
+    .groupBy(renderParams)
+    .values()
+    .map(values => chunk(values, 20))
+    .flatten()
+    .map((keys: string[] | { id }[]) => {
+      if (typeof keys[0] === "string") {
+        return { id: keys, size: keys.length }
+      }
+      return {
+        // @ts-ignore
+        ...keys[0],
+        id: keys.map(k => k.id),
+        size: keys.length,
+      }
+    })
+    .value() as unknown) as GroupKeysResult[]
+
+interface BatchLoaderArgs {
+  /**
+   * For resolvers designed to hit a list endpoint
+   * leave this argument empty. List endpoints
+   * support filtering options that single resource
+   * endpoints do not.
+   */
+  singleLoader?: any
+  multipleLoader: any
+  defaultResult?: any
+}
+
+export const batchLoader = ({
+  singleLoader,
+  multipleLoader,
+  defaultResult = null,
+}: BatchLoaderArgs) => {
+  if (!ENABLE_RESOLVER_BATCHING) {
+    return singleLoader ? singleLoader : multipleLoader
+  }
+  const dl = new DataLoader(keys => {
+    let groupedKeys = groupKeys(keys as any)
+
+    return Promise.all(
+      groupedKeys.map(keys => {
+        if (keys.id.length === 1 && singleLoader) {
+          return singleLoader(keys.id[0])
```

Comment:
The thing that gets passed into the single loader is always just the string id. That seems to hold true for the sale endpoint (there aren't other params that I've seen), but that might not hold true for every endpoint.